chore: add ClayAutocomplete component docs and update gatsby-mdx (#2070)
liferay/clay: next.clayui.com/content/docs/components/autocomplete.mdx

packageNpm: "@clayui/autocomplete"
packageVersion: "3.0.0"
packageStatus: "Beta"
---

import {AutocompleteWithLoadingState, AutocompleteWithDataProvider} from '../../../src/components/clay/Autocomplete';

<div class="nav-toc">

- [Item](#item)
- [Loading state](#loading-state)
- [Data from outside](#data-from-outside)
- [API](#api)
  - [Autocomplete](#autocomplete)
  - [Autocomplete.DropDown](#autocomplete.dropdown)
  - [Autocomplete.Input](#autocomplete.input)
  - [Autocomplete.Item](#autocomplete.item)
  - [Autocomplete.LoadingIndicator](#autocomplete.loadingindicator)

</div>

<div class="clay-site-alert alert alert-info">
	Check the <a href="https://liferay.design/lexicon/">Lexicon</a> <a href="https://liferay.design/lexicon/core-components/forms/text-input-variations/">Autocomplete Pattern</a> for a more in-depth look at the motivations and proper usage of this component.
</div>

Autocomplete provides the necessary components for you to compose with the other components, **by itself it does not do much**, you need to use the `<ClayDropDown />` along with `<Autocomplete.DropDown />` and `<Autocomplete.Item />` to create the suggestions.

```jsx
<ClayAutocomplete.DropDown>

> 	<ClayDropDown.ItemList>

Author: I'd rather keep separate, I think export like `<Autocomplete.ItemList />` takes away the beauty of seeing the composition between components and how they can relate. And adding it to `<ClayAutocomplete.DropDown />` can take away other customization possibilities. For example, I can add headers to my list. Probably I could create several `<ClayDropDown.ItemList />` inside the `<ClayAutocomplete.DropDown>`.

> 	<ClayDropDown.ItemList>

Reviewer: The main concern I have is that `@clayui/dropdown` is now a peer dependency of autocomplete since we now always expect the user to use it when they render an autocomplete list. Really the user doesn't need to know that Autocomplete is the composition of using dropdown, the user just wants to use the component for its intended purpose.

> 	<ClayDropDown.ItemList>

Author: Yes we can agree, we will have people who just want to use the component without needing to know that it needs to compose things, probably these people are the great candidates to use the high-level components, probably those who want to create their own autocomplete with their rules, will likely compose with other components.

> 	<ClayDropDown.ItemList>

Reviewer: Yeah makes sense, if we do keep this API by keeping dropdown separate, we will need a peerDependency of dropdown right?

> 	<ClayDropDown.ItemList>

Author: Maybe, I think we should change all the dependencies we have between the components for `peerDependency`, currently they are like` dependencies` the dropdown also in that case.

		<ClayAutocomplete.Item
			match="cl"
			value="clay"
		/>
	</ClayDropDown.ItemList>
</ClayAutocomplete.DropDown>
```

When you already have the raw data and do not need to consult them, it becomes simpler to create an autocomplete. It is important that you **create a filter of your data** according to the value entered in the input element, following your business rule.

```jsx
<ClayAutocomplete>
	<ClayAutocomplete.Input
		onChange={event => setValue(event.target.value)}
		value={value}
	/>
	<ClayAutocomplete.DropDown active={showDropDown}>
		<ClayDropDown.ItemList>
			{resourceFiltered.map(item => (
				<ClayAutocomplete.Item
					key={item}
					match={value}
					value={item}
				/>
			))}
		</ClayDropDown.ItemList>
	</ClayAutocomplete.DropDown>
</ClayAutocomplete>
```

> ## Item

Author: Hm, makes sense.

`ClayAutocomplete.Item` is a wrapper above `ClayDropDown.Item` but covers the use case to underline the bold characters in case you follow the Lexicon. Feel free to create your own match rule using `ClayDropDown.Item` which will give you more freedom to do so.

<div class="sheet-example">
	<ul class="list-unstyled">
		<li>
			<span class="dropdown-item">
				<div>
					<strong>C</strong>
					<strong>l</strong>
					ay
				</div>
			</span>
		</li>
	</ul>
</div>

```jsx
<ClayAutocomplete.Item
	match="Cl"
	value="Clay"
/>
```

## Loading state

> `<ClayAutocomplete.LoadingIndicator />` is used for cases where you need to provide the user with feedback that is loading new data to fill in the suggestion box, it does communicate with `<ClayAutocomplete.Input />` for the markup can adjust.

Reviewer: ```suggestion
`<ClayAutocomplete.LoadingIndicator />` is used for providing feedback that data is loading in the suggestion box, it communicates with `<ClayAutocomplete.Input />` so that the markup can adjust.
```

<div class="clay-site-alert alert alert-warning">
	<strong class="lead">Warning</strong>
	Add this when your suggestion data is coming from a data provider.
</div>

<AutocompleteWithLoadingState />

## Data from outside

> For cases where you do not have all the data in the client and need to pass the value for the filter to the backend, use the data provider for this, `<ClayDataProvider />` component or `useResource` hook.

Reviewer: ```suggestion
For cases where you do not have all the data in the client and need to pass the value for the filter to the backend, use the `<ClayDataProvider />` component or `useResource` hook for this, .
```
